pkg/ui/src/views/shared/components/sortedtable/index.tsx

```diff
@@ -45,6 +50,22 @@ interface SortedTableProps<T> {
   className?: string;
   // A function that returns the class to apply to a given row.
   rowClass?: (obj: T) => string;
+
+  // expandableConfig, if provided, makes each row in the table "expandable",
+  // i.e. each row has an expand/collapse arrow on its left, and renders
+  // a full-width area below it when expanded.
+  expandableConfig?: {
+    // expandedContent returns the content for a row's full-width expanded
+    // section, given the object that row represents.
+    expandedContent: (obj: T) => React.ReactNode;
+    // expansionKey returns a key used to uniquely identify a row for the
+    // purposes of tracking whether it's expanded or not.
+    expansionKey: (obj: T) => string;
+  };
+}
+
+interface SortedTableState {
+  expandedRows: Set<string>;
```

Comment:
It looks like the changes at this level are mostly about keeping state for the sortable table below.  The pattern to separate state-keeping containers from stateless components makes sense, but I'm not sure if the state belongs here.  In this case it splits the logic and data structures (there are two `expandableConfig` props with similar but different keys, for instance).  It seems like it would be clearer either inline in sortable table itself or in a thin wrapper in the same file.

Reply:
The `expandableConfig` thing is annoying, but it still seems to make sense to keep the expansion state next to the sort state, in `SortedTable`. Not quite following your suggestion.